[breaking] error handling rework (#6586)
sveltejs/kit: documentation/docs/07-hooks.md

title: Hooks
---

For both the client and the server, a range of app-wide hooks can be specified.

### hooks.client.js

An optional `src/hooks.client.js` (or `src/hooks.client.ts`, or `src/hooks.client/index.js`) file exports one functions, optional, that run on the client — `handleError`.

> The location of this file can be [configured](/docs/configuration) as `config.kit.files.hooks.server`

#### handleError

If an error is thrown during loading or rendering, this function will be called with the `error`. This allows for two things:

- you can send data to an error tracking service, for example

> - you can customise the formatting. The return value of the function becomes the value of `$page.error`. To make this type-safe, you can customize the expected shape through `App.PageError`. Note that `message` is always a required property, to guarantee a sensible fallback behavior.

Reviewer: ```suggestion
- you can customise the formatting. The return value of the function becomes the value of `$page.error`. To make this type-safe, you can customize the expected shape through `App.PageError`, which must include a `message: string` property.
```

Author: I liked the previous version better because it also explains why.

```js
/// file: src/hooks.client.js
// @filename: ambient.d.ts
const Sentry: any;

// @filename: index.js
// ---cut---
/** @type {import('@sveltejs/kit').HandleClientError} */
export function handleError({ error }) {
	// example integration with https://sentry.io/
	Sentry.captureException(error);
	return { message: /** @type {Error} */ (error).message }
}
```

> If unimplemented or no error shape is returned, SvelteKit will use the default formatting `{ message: string }`.

Reviewer: ```suggestion
The default return value, if none is provided or `handleError` is not implemented, is `{ message: 'Internal Error' }`.
```

}
```

#### handleError

If an error is thrown during loading or rendering, this function will be called with the `error` and the `event` that caused it. This allows for two things:

- you can send data to an error tracking service, for example

> - you can customise the formatting. The return value of the function becomes the value of `$page.error`. To make this type-safe, you can customize the expected shape through `App.PageError`. Note that `message` is always a required property, to guarantee a sensible fallback behavior.

Reviewer: ```suggestion
- you can customise the formatting. The return value of the function becomes the value of `$page.error`. To make this type-safe, you can customize the expected shape through `App.PageError`, which must include a `message: string` property.
```

Author: I liked the previous version better because it also explains why.

```js
/// file: src/hooks.server.js
// @filename: ambient.d.ts
const Sentry: any;

// @filename: index.js
// ---cut---
/** @type {import('@sveltejs/kit').HandleServerError} */
export function handleError({ error, event }) {
	// example integration with https://sentry.io/
	Sentry.captureException(error, { event });
	return { message: /** @type {Error} */(error).message }
}
```

> If unimplemented or no error shape is returned, SvelteKit use the default formatting `{ message: string }`.

Reviewer: ```suggestion
The default return value, if none is provided or `handleError` is not implemented, is `{ message: 'Internal Error' }`.
```

> During development, if an error occurs because of a syntax error in your Svelte code, a `frame` property will be appended highlighting the location of the error.

Reviewer: Is this still true? Feels like this stuff should be kept out of the error object

```suggestion
```

Author: Mhm not sure, but I feel like it should (hopefully?) still be true - only that it's passed in, but not necessarily appears in your user-facing error page.

Reviewer: Ah right, I mistook input for output. Sleep deprivation
